packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m

```diff
@@ -0,0 +1,346 @@
+// Copyright 2018-present 650 Industries. All rights reserved.
+
+#import <EXInAppPurchases/EXInAppPurchasesModule.h>
+
+#define QUERY_HISTORY_KEY @"QUERY_HISTORY"
+#define QUERY_PURCHASABLE_KEY @"QUERY_PURCHASABLE"
+#define SERVICE_DISCONNECTED -1
+#define OK 0
+#define ERROR 1
+#define USER_CANCELED 2
+
+@implementation EXInAppPurchasesModule
+
+UM_EXPORT_MODULE(ExpoInAppPurchases);
+
+- (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
+{
+  _moduleRegistry = moduleRegistry;
+}
+
+- (NSDictionary *)constantsToExport
+{
+  return @{
+       @"responseCodes": @{
+         @"OK": @OK,
+         @"USER_CANCELED": @USER_CANCELED,
+         @"ERROR": @ERROR,
+         @"SERVICE_DISCONNECTED": @SERVICE_DISCONNECTED
```

Comment:
Can we move all these constants to TypeScript's enums? We should rather keep the consistent (and typed) API on the JS side and let the native side translate that JS values to platform-specific values. 

```typescript
export enum ResponseCodes {
  Success = 0,
  Canceled = 1,
  Error = 2,
  Disconnected = 3,
  /* at then some other Android-specific response codes */
};
```

Reply:
That would be a good idea, however the caveat here is that iOS and Android return different codes for the same thing.

For instance, if a user cancels a payment Android returns [USER_CANCELED](https://developer.android.com/reference/com/android/billingclient/api/BillingClient.BillingResponseCode.html#USER_CANCELED) (enum value 1), but iOS will return [SKErrorPaymentCancelled](https://developer.apple.com/documentation/storekit/skerror/code/paymentcancelled) which is an enum value for 2. They even spell "Canceled" differently which I discovered the hard way 😄

Likewise, a successful purchase returns "OK" (enum value 0) on Android and "SKPaymentTransactionStatePurchased" (enum value 1) on iOS. That's why I wanted to return different values on the Native side just be consistent with the keys I use so that it surfaces the same on the TS layer.

Reply:
Yeah, I know these native values could be different, however we can translate them on the native side. Take a look at the `LocationAccuracy` in `expo-location`:

TypeScript:
https://github.com/expo/expo/blob/6cd194c7469bf80f91f4f819b15e555538cf3479/packages/expo-location/src/Location.ts#L94-L101

Java:
https://github.com/expo/expo/blob/6cd194c7469bf80f91f4f819b15e555538cf3479/packages/expo-location/android/src/main/java/expo/modules/location/LocationHelpers.java#L200-L234

On Android there are just 4 accuracy levels, so we translate the TypeScript ones to the platform-specific one. In other words, native side can map TypeScript values to the ones used by the platform. Another good thing is that we wouldn't need to export constants by the module - it's an additional data that we need to send through the bridge while initializing modules.

Reply:
Yeah, I was thinking about just translating the iOS constants to match those on Android to reduce the need to export them at all.

However, in the payments code I also have this logic for when a purchase fails:
```objective-c
NSString *errorCode = [NSString stringWithFormat:@"%ld", (long)transaction.error.code];
[self rejectPromise:transaction.payment.productIdentifier code:errorCode message:transaction.error.localizedDescription error:transaction.error];
```

And if you take a look at all of the possible error codes that can be returned, they actually conflict with those that Android returns. So for example, if I standardize that 0 is a successful purchase (as it is on Android) and make that a TS enum, a developer might get confused if their purchase promise on iOS rejects with error code 0 since that is also the code for [SKErrorUnknown](https://developer.apple.com/documentation/storekit/skerror/code/unknown)

This is an issue with other values as well, and I want to make sure I'm returning the correct error code on iOS since that is useful information for a developer.

Reply:
So James and I discussed this and came up with a new solution that allows for a TS enum. See this commit: eec206a

This loses a little information but handles the issue of overlap with error codes and response codes and converts the results natively to match new TS enums. Let me know what you think.